waf.go

```diff
@@ -3,975 +3,44 @@
-// size that was allocated and copied.
-func cstring(str string, maxLength int) (*C.char, int, error) {
-	// Limit the maximum string size to copy
-	l := len(str)
-	if l > maxLength {
-		l = maxLength
-	}
-	// Copy the string up to l.
-	// The copy is required as the pointer will be stored into the C structures,
-	// so using a Go pointer is impossible.
-	cstr := C.CString(str[:l])
-	if cstr == nil {
-		return nil, 0, errOutOfMemory
-	}
-	return cstr, l, nil
-}
-
-func freeWO(v *wafObject) {
-	if v == nil {
-		return
-	}
-	// Free the map key if any
-	if key := v.mapKey(); key != nil {
-		cFree(unsafe.Pointer(v.parameterName))
-	}
-	// Free allocated values
-	switch v._type {
-	case wafInvalidType:
-		return
-	case wafStringType:
-		cFree(unsafe.Pointer(v.string()))
-	case wafMapType, wafArrayType:
-		freeWOContainer(v)
-	}
-	// Make the value invalid to make it unusable
-	v.setInvalid()
-}
-
-func freeWOContainer(v *wafObject) {
-	length := v.length()
-	for i := C.uint64_t(0); i < length; i++ {
-		freeWO(v.index(i))
-	}
-	if a := *v.arrayValuePtr(); a != nil {
-		cFree(unsafe.Pointer(a))
-	}
-}
-
-func cFree(ptr unsafe.Pointer) {
-	C.free(ptr)
-}
-
-// Go reimplementation of ddwaf_result_free to avoid yet another CGO call in the
-// request hot-path and avoiding it when there are no results to free.
-func freeWAFResult(result *C.ddwaf_result) {
-	if data := result.data; data != nil {
-		C.free(unsafe.Pointer(data))
-	}
-
-	if array := result.actions.array; array != nil {
-		for i := 0; i < int(result.actions.size); i++ {
-			C.free(unsafe.Pointer(cindexCharPtrArray(array, i)))
-		}
-		C.free(unsafe.Pointer(array))
-	}
-}
-
-// Helper function to access to i-th element of the given **C.char array.
-func cindexCharPtrArray(array **C.char, i int) *C.char {
-	// Go pointer arithmetic equivalent to the C expression `array[i]`
-	base := uintptr(unsafe.Pointer(array))
-	return *(**C.char)(unsafe.Pointer(base + unsafe.Sizeof((*C.char)(nil))*uintptr(i))) //nolint:govet
-}
-
-// Atomic reference counter helper initialized at 1 so that 0 is the special
-// value meaning that the object was released and is no longer usable.
-type atomicRefCounter atomic.Uint32
-
-func (c *atomicRefCounter) unwrap() *atomic.Uint32 {
-	return (*atomic.Uint32)(c)
-}
-
-// Add delta to reference counter.
-// It relies on a CAS spin-loop implementation in order to avoid changing the
-// counter when 0 has been reached.
-func (c *atomicRefCounter) add(delta uint32) uint32 {
-	addr := c.unwrap()
-	for {
-		current := addr.Load()
-		if current == 0 {
-			// The object was released
-			return 0
-		}
-		new := current + delta
-		if swapped := addr.CompareAndSwap(current, new); swapped {
-			return new
-		}
-	}
-}
-
-// Increment the reference counter.
-// CAS spin-loop implementation in order to enforce +1 cannot happen when 0 has
-// been reached.
-func (c *atomicRefCounter) increment() uint32 {
-	return c.add(1)
+func Health() error {
+	return InitWaf()
```

Comment:
As of the current state of `InitWaf`, `Health()` can currently fail if things like `os.Remove` fails, while we must be more permissive and let using the libddwaf as long as it is usable. OTOH, I think all the errors are worth noting via the dd-trace-go logger, which we can't access from here.

For this reason, I suggest we move to a slightly different API where we don't loose the details:
1. Rename `InitWaf()` into `Load()`, and `CloseWaf()` into `Close()`.
2. Change the API of `Load()` into `func Load() (ok bool, err error)` so that we can let the caller can know if libddwaf is usable or not thanks to `ok`:
    1. When `ok` is true, a non-nil error is acceptable to me in order to return such `os.Remove()` errors for instance that we can log in the dd-trace-go side.
    2. When `ok` is false (ie. not usable), the error must state why, and so such `os.Remove()` errors shouldn't take precedence over the actual libddwaf-related errors.

I'm ok doing such changes in a dedicated separate PR.

Reply:
Sure let there be a Jira ticket

Reply:
done